include/psa/crypto.h

```diff
@@ -4043,6 +4043,553 @@ psa_status_t psa_generate_random(uint8_t *output,
 psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
                               mbedtls_svc_key_id_t *key);
 
+/**@}*/
+
+/** \defgroup interruptible_hash Interruptible sign/verify hash
+ * @{
+ */
+
+/** The type of the state data structure for interruptible hash
+ *  verification operations.
+ *
+ * Before calling any function on a sign hash operation object, the
+ * application must initialize it by any of the following means:
+ * - Set the structure to all-bits-zero, for example:
+ *   \code
+ *   psa_sign_operation_t operation;
+ *   memset(&operation, 0, sizeof(operation));
+ *   \endcode
+ * - Initialize the structure to logical zero values, for example:
+ *   \code
+ *   psa_sign_operation_t operation = {0};
+ *   \endcode
+ * - Initialize the structure to the initializer
+ *   #PSA_SIGN_HASH_OPERATION_INIT, for example:
+ *   \code
+ *   psa_sign_operation_t operation = PSA_SIGN_HASH_OPERATION_INIT;
+ *   \endcode
+ * - Assign the result of the function psa_sign_hash_operation_init() to
+ *   the structure, for example:
+ *   \code
+ *   psa_sign_operation_t operation;
+ *   operation = psa_sign_hash_operation_init();
+ *   \endcode
+ *
+ * This is an implementation-defined \c struct. Applications should not
+ * make any assumptions about the content of this structure.
+ * Implementation details can change in future versions without notice. */
+typedef struct psa_sign_operation_s psa_sign_hash_operation_t;
+
+/** The type of the state data structure for interruptible hash signing
+ *  operations.
+ *
+ * Before calling any function on a sign hash operation object, the
+ * application must initialize it by any of the following means:
+ * - Set the structure to all-bits-zero, for example:
+ *   \code
+ *   psa_verify_operation_t operation;
+ *   memset(&operation, 0, sizeof(operation));
+ *   \endcode
+ * - Initialize the structure to logical zero values, for example:
+ *   \code
+ *   psa_verify_operation_t operation = {0};
+ *   \endcode
+ * - Initialize the structure to the initializer
+ *   #PSA_VERIFY_HASH_OPERATION_INIT, for example:
+ *   \code
+ *   psa_verify_operation_t operation = PSA_VERIFY_HASH_OPERATION_INIT;
+ *   \endcode
+ * - Assign the result of the function psa_verify_hash_operation_init() to
+ *   the structure, for example:
+ *   \code
+ *   psa_verify_operation_t operation;
+ *   operation = psa_verify_hash_operation_init();
+ *   \endcode
+ *
+ * This is an implementation-defined \c struct. Applications should not
+ * make any assumptions about the content of this structure.
+ * Implementation details can change in future versions without notice. */
+typedef struct psa_verify_operation_s psa_verify_hash_operation_t;
+
+/**
+ * \brief                       Set the maximum number of basic operations
```

Comment:
This is a “maximum number”. So on a hard real time system, this can indicate a hard deadline, which is good: a real-time system needs a guarantee of not spending more than X time. On a non-real-time system, should the guidance be to do this many ops or less? Or to do about this many ops, possibly more?

However. If the “max” is too small, a call to complete() might do not work at all, resulting in an infinite loop with no progress. How do we balance the needs of hard real-time systems with the need to make progress?

Reply:
I have added warnings here that the maximum number could be interpeted as a hard limit by some implementations, and thus setting the number too small could cause infinate loop with no result. My only other suggestion is to add (another) return code that says 'everything ok, however I didn't do any work'

Reply:
I think it would be reasonable to say that implementations may have some lower threshold for the size of a step, below which reducing the maximum has no effect (i.e. there will always be some progress). If a hard-realtime system needs the threshold so low it's not possible to make progress, then it's broken either way.